src/PBTRandom.sol

```diff
@@ -0,0 +1,272 @@
+
+        emit PBTMint(tokenId, chipAddr);
+
+        return tokenId;
+    }
+
+    // Generates a pseudorandom number between [0,maxSupply) that has not yet been generated before, in O(1) time.
+    //
+    // Uses Durstenfeld's version of the Yates Shuffle https://en.wikipedia.org/wiki/Fisher%E2%80%93Yates_shuffle
+    // with a twist to avoid having to manually spend gas to preset an array's values to be values 0...n.
+    // It does this by interpreting zero-values for an index X as meaning that index X itself is an available value
+    // that is returnable.
+    //
+    // How it works:
+    //  - zero-initialize a mapping (_availableRemainingTokens) and track its length (_numAvailableRemainingTokens). functionally similar to an array with dynamic sizing
+    //    - this mapping will track all remaining valid values that haven't been generated yet, through a combination of its indices and values
+    //      - if _availableRemainingTokens[x] == 0, that means x has not been generated yet
+    //      - if _availableRemainingTokens[x] != 0, that means _availableRemainingTokens[x] has not been generated yet
+    //  - when prompted for a random number between [0,maxSupply) that hasn't already been used:
+    //    - generate a random index randIndex between [0,_numAvailableRemainingTokens)
+    //    - examine the value at _availableRemainingTokens[randIndex]
+    //        - if the value is zero, it means randIndex has not been used, so we can return randIndex
+    //        - if the value is non-zero, it means the value has not been used, so we can return _availableRemainingTokens[randIndex]
+    //    - update the _availableRemainingTokens mapping state
+    //        - set _availableRemainingTokens[randIndex] to either the index or the value of the last entry in the mapping (depends on the last entry's state)
+    //        - decrement _numAvailableRemainingTokens to mimic the shrinking of an array
+    function _useRandomAvailableTokenId() internal returns (uint128) {
+        uint128 numAvailableRemainingTokens = _numAvailableRemainingTokens;
+
+        // Devs can swap this out for something less gameable like chainlink if it makes sense for their use case.
+        uint256 randomNum = uint256(
+            keccak256(
+                abi.encode(
+                    _msgSender(),
+                    tx.gasprice,
+                    block.number,
+                    block.timestamp,
+                    block.difficulty,
+                    blockhash(block.number - 1),
+                    address(this),
+                    numAvailableRemainingTokens
+                )
+            )
+        );
+        uint128 randomIndex = uint128(randomNum % numAvailableRemainingTokens);
+        return _setAvailableTokenAtIndex(randomIndex);
+    }
+
+    function _setAvailableTokenAtIndex(uint128 indexToUse) internal returns (uint128) {
+        uint128 numAvailableRemainingTokens = _numAvailableRemainingTokens;
```

Comment:
Note: I removed this as an input argument to the function. I know it was added in to save gas, but post Berlin merge, SLOADs to accessed storage slots only cost 100 gas. It makes the code a lot easier to read and test for now. I think it could be worth it to merge these this function and the one above, but open to suggestions.

https://hackmd.io/@fvictorio/gas-costs-after-berlin#Putting-it-all-together

Reply:
Also think it's worth merging these functions together since I can't think of any valid reason why a dev would call `_getAvailableTokenAtIndex` through a different call path